maestro-client/src/main/java/maestro/js/GraalJsEngine.kt

```diff
@@ -0,0 +1,115 @@
+package maestro.js
+
+import okhttp3.OkHttpClient
+import okhttp3.Protocol
+import org.graalvm.polyglot.Context
+import org.graalvm.polyglot.Source
+import org.graalvm.polyglot.Value
+import org.graalvm.polyglot.proxy.ProxyObject
+import java.io.ByteArrayOutputStream
+import java.util.concurrent.TimeUnit
+import java.util.logging.Handler
+import java.util.logging.LogRecord
+
+private val NULL_HANDLER = object : Handler() {
+    override fun publish(record: LogRecord?) {}
+
+    override fun flush() {}
+
+    override fun close() {}
+}
+
+class GraalJsEngine(
+    httpClient: OkHttpClient = OkHttpClient.Builder()
+        .readTimeout(5, TimeUnit.MINUTES)
+        .writeTimeout(5, TimeUnit.MINUTES)
+        .protocols(listOf(Protocol.HTTP_1_1))
+        .build(),
+) : JsEngine {
+
+    private val openContexts = HashSet<Context>()
+
+    private val httpBinding = GraalJsHttp(httpClient)
+    private val outputBinding = HashMap<String, Any>()
+    private val maestroBinding = HashMap<String, Any?>()
+    private val envBinding = HashMap<String, String>()
+
+    private var onLogMessage: (String) -> Unit = {}
+
+    override fun close() {
+        openContexts.forEach { it.close() }
+    }
+
+    override fun onLogMessage(callback: (String) -> Unit) {
+        onLogMessage = callback
+    }
+
+    override fun enterScope() {}
+
+    override fun leaveScope() {}
+
+    override fun putEnv(key: String, value: String) {
+        this.envBinding[key] = value
+    }
+
+    override fun setCopiedText(text: String?) {
+        this.maestroBinding["copiedText"] = text
+    }
+
+    override fun evaluateScript(
+        script: String,
+        env: Map<String, String>,
+        sourceName: String,
+        runInSubScope: Boolean,
+    ): Value {
+        envBinding.putAll(env)
+        val source = Source.newBuilder("js", script, sourceName).build()
+        return createContext().eval(source)
+    }
+
+    private fun createContext(): Context {
+        val outputStream = object : ByteArrayOutputStream() {
+            override fun flush() {
+                super.flush()
+                val log = toByteArray().decodeToString().removeSuffix("\n")
+                onLogMessage(log)
+                reset()
+            }
+        }
+
+        val context = Context.newBuilder("js")
+            .option("js.strict", "true")
+            .logHandler(NULL_HANDLER)
+            .out(outputStream)
+            .build()
+
+        openContexts.add(context)
+
+        envBinding.forEach { (key, value) -> context.getBindings("js").putMember(key, value) }
+
+        context.getBindings("js").putMember("http", httpBinding)
+        context.getBindings("js").putMember("output", ProxyObject.fromMap(outputBinding))
```

Comment:
`output` and `maestro` are the two variables (aka "bindings") shared across scripts